internal/typeinfo/typeinfo.go

```diff
@@ -0,0 +1,105 @@
+package typeinfo
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+)
+
+var cacheMutex sync.RWMutex
+var cache = make(map[reflect.Type]Info)
+
+// Reflect will return the Info of a given type,
+// generating and caching as required.
+func GetTypeInfo(value any) (Info, error) {
+	if value == (any)(nil) {
+		return Info{}, fmt.Errorf("Can not reflect nil value")
+	}
+
+	v := reflect.ValueOf(value)
+	v = reflect.Indirect(v)
+
+	cacheMutex.RLock()
+	info, found := cache[v.Type()]
+	cacheMutex.RUnlock()
+	if found {
+		return info, nil
+	}
+
+	info, err := generate(v)
+	if err != nil {
+		return Info{}, err
+	}
+
+	cacheMutex.Lock()
+	cache[v.Type()] = info
+	cacheMutex.Unlock()
+
+	return info, nil
+}
+
+// generate produces and returns reflection information for the input
+// reflect.Value that is specifically required for SQLAir operation.
+func generate(value reflect.Value) (Info, error) {
+	// Dereference the pointer if it is one.
+	value = reflect.Indirect(value)
+
+	// Reflection information is generated for structs, typeinfo.M
+	// and plain types only.
+	if value.Kind() != reflect.Struct {
+		if value.Kind() == reflect.Map && value.Type().Name() != "M" {
+			return Info{}, fmt.Errorf("Can't reflect map type")
+		} else {
+			return Info{Type: value.Type()}, nil
```

Comment:
Per prior review on this PR, I don't think we should be handling maps here, M or otherwise.  This functionality is about extracting tag information from structs, so that we can tell which fields map to which columns and vice versa. The map handling by definition needs to be done in a very different way. If we use the struct type info in a map, we have a bug.

Reply:
Done. This code has been removed.